sky/backends/wheel_utils.py

```diff
@@ -5,82 +5,104 @@
+                                f'under {WHEEL_DIR!r}') from None
+
+    latest_wheel_dir_name = latest_wheel.parent
     # Cleanup older wheels.
-    for f in wheel_dir.iterdir():
-        if f != latest_wheel:
+    for f in WHEEL_DIR.iterdir():
+        if f != latest_wheel_dir_name:
             if f.is_dir() and not f.is_symlink():
                 shutil.rmtree(f, ignore_errors=True)
             else:
                 f.unlink()
+    return latest_wheel
 
 
-def _get_latest_built_wheel() -> pathlib.Path:
-    wheel_name = (f'{_PACKAGE_WHEEL_NAME}-'
-                  f'{version.parse(sky.__version__)}-*.whl')
+def clean_up_wheel_dir():
+    """Clean up the wheel directory.
+
+    This function will be called during `ray up` to clean up the wheel, it
+    acquires the wheel lock to prevent concurrent wheel operations, by other
+    `ray up` or `sky launch`.
+    """
     try:
-        latest_wheel = max(WHEEL_DIR.glob(wheel_name), key=os.path.getctime)
-    except ValueError:
-        raise FileNotFoundError(
-            f'Could not find built SkyPilot wheels {wheel_name!r} '
-            f'under {WHEEL_DIR!r}') from None
-    return latest_wheel
+        with filelock.FileLock(_WHEEL_LOCK_PATH, timeout=5):  # pylint: disable=E0110
+            _get_latest_wheel_and_cleanup()
+    except FileNotFoundError as e:
+        logger.error(str(e))
+    except filelock.Timeout:
+        logger.warning('Failed to clean up wheel directory')
 
 
-def _build_sky_wheel() -> pathlib.Path:
+def _build_sky_wheel():
     """Build a wheel for Sky."""
-    # prepare files
-    if (WHEEL_DIR / 'sky').exists():
-        # This may because the last wheel build was interrupted,
-        # so the symlink file doesn't get cleaned.
-        (WHEEL_DIR / 'sky').unlink()
-    (WHEEL_DIR / 'sky').symlink_to(SKY_PACKAGE_PATH, target_is_directory=True)
-    setup_files_dir = SKY_PACKAGE_PATH / 'setup_files'
-    for f in setup_files_dir.iterdir():
-        if f.is_file():
-            shutil.copy(str(f), str(WHEEL_DIR))
-
-    # It is important to normalize the path, otherwise 'pip wheel' would
-    # treat the directory as a file and generate an empty wheel.
-    norm_path = str(WHEEL_DIR) + os.sep
-    try:
-        # TODO(suquark): For python>=3.7, 'subprocess.run' supports capture
-        # of the output.
-        subprocess.run([
-            'pip3', 'wheel', '--no-deps', norm_path, '--wheel-dir',
-            str(WHEEL_DIR)
-        ],
-                       stdout=subprocess.DEVNULL,
-                       stderr=subprocess.PIPE,
-                       check=True)
-    except subprocess.CalledProcessError as e:
-        raise RuntimeError('Fail to build pip wheel for Sky. '
-                           f'Error message: {e.stderr.decode()}') from e
-
-    latest_wheel = _get_latest_built_wheel()
-    cleanup_wheels_dir(WHEEL_DIR, latest_wheel)
-    return WHEEL_DIR.absolute()
+    with tempfile.TemporaryDirectory() as tmp_dir:
```

Comment:
Q: Why use a temp dir, since the caller already holds a lock?

Reply:
The reason is that we can only compute the hash for the wheel after the wheel is generated, ie we do not know the directory name during building.